fedimint-server/src/consensus/engine.rs

```diff
@@ -584,13 +604,118 @@ impl ConsensusEngine {
         Ok(())
     }
 
+    async fn process_consensus_version_vote(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        peer_id: PeerId,
+        vote: ConsensusVersionVote,
+    ) -> Result<(), anyhow::Error> {
+        let key = ConsensusVersionVoteKey::from((vote, peer_id));
+        let val = ConsensusVersionVoteValue::from(vote);
+
+        if let Some(module_id) = key.module_id {
+            if !self.cfg.consensus.modules.contains_key(&module_id) {
+                warn!(target: LOG_CONSENSUS, %peer_id, module_id, ?key, ?val, "Peer submitted invalid consensus version vote: invalid module id");
+                bail!("Invalid vote: invalid module id");
+            }
+        }
+        if Some(val) == dbtx.get_value(&key).await {
+            warn!(target: LOG_CONSENSUS, %peer_id, ?key, ?val, "Peer submitted duplicate consensus version vote");
+            bail!("Redundant vote");
+        }
+        dbtx.insert_entry(&key, &val).await;
+
+        if let Some(desired_version) = self
+            .get_desired_consensus_version(dbtx, key.module_id)
+            .await
+        {
+            info!(target: LOG_CONSENSUS, module_id=?key.module_id, version=%desired_version, "Updating consensus value");
+            dbtx.insert_entry(
+                &ConsensusVersionKey(key.module_id),
+                &ConsensusVersionValue {
+                    major: desired_version.major,
+                    minor: desired_version.minor,
+                },
+            )
+            .await;
+        }
+
+        Ok(())
+    }
+
+    fn get_desired_consensus_version_from_votes_static(
+        num_peers: NumPeers,
+        votes: impl IntoIterator<Item = ConsensusVersionVoteValue>,
+    ) -> Option<ConsensusVersionValue> {
+        let mut votes_by_major: BTreeMap<u32, BTreeMap<u32, usize>> = Default::default();
+        for ConsensusVersionVoteValue { major, minor } in votes {
+            *votes_by_major
+                .entry(major)
+                .or_default()
+                .entry(minor)
+                .or_default() += 1;
+        }
+
+        let threshold_needed = num_peers.threshold();
+
+        for (major, minor_counts) in votes_by_major {
+            let mut threshold_accumulated = 0;
+            for (minor, count) in minor_counts.into_iter().rev() {
+                threshold_accumulated += count;
+
+                if threshold_needed <= threshold_accumulated {
+                    return Some(ConsensusVersionValue { major, minor });
+                }
+            }
+        }
+
+        None
+    }
```

Comment:
>  Ideally everyone runs the same software version

I do not understand. This code needs to support switching between any versions, including major version switches if we ever have many, for the purpose of rewinding consensus from any past versions.

20 LoC is not much, IMO, and I moved it to a separate function so testing it thoroughly

Reply:
Hmmm... Or should we just use major version to mean `v2` in `lnv2`, and not support ever upgrading major consensus version in the running federations?

Reply:
I was thinking that you should never have a situation where you run versions `1.0`, `1.1`, `1.2` and `1.1` together. If an upgrade happens that should work as follows:

At first everyone is at `1.0`, then everyone upgrades to e.g. `1.1` and once enough updated they vote on `1.1`. In that case it would be sufficient to check which version, if any, has enough votes to be activated, otherwise stick with the previous one.

Maybe that's too restrictive, but I doubt we'll really support running arbitrary code versions with each other, it's just too much testing effort. It's less about this code and more about "do we want to support this at all?".